### PR TITLE
Spider for gulahmedshop.com

### PR DESCRIPTION
**Pull request by mateen91**

### CODE REVIEW

File: gulahmed/gulahmed/spiders/gul_products.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+
+class MySpider(CrawlSpider):
+    name = "gulahmed"
+    allowed_domains = ["gulahmedshop.com"]
+    start_urls = ['http://www.gulahmedshop.com/']
+
+    rules = (
+        Rule(SgmlLinkExtractor(restrict_xpaths=("//*[@id='site-menu']/ul/li//li",
+                                                ))),
```
Comment (ali-ashraf-karimi) [inline gulahmed/gulahmed/spiders/gul_products.py]: Try this: //*[@id='site-menu']//ul[@class="mega-sub"]

File: gulahmed/gulahmed/spiders/gul_products.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        Rule(SgmlLinkExtractor(restrict_xpaths=("//*[@id='main-content']//section/section[1]//*[@id='brands-list']",
+                                               )),
+            callback="parse_items")
+    )
+
+    # this will parse all the products on current page
+    def parse_items(self, response):
+
+        sel = response.xpath("/html")
+        nextpage = sel.xpath("//*[@id='frm']//a[@data-title='Next Page']/@href")
```
Comment (ali-ashraf-karimi) [inline gulahmed/gulahmed/spiders/gul_products.py]: Try this: //a[@data-title='Next Page']/@href

File: gulahmed/gulahmed/spiders/gul_products.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+            yield scrapy.Request(next_page_url, callback=self.parse_items)
+
+        # getting all products on current page
+        product_links = response.xpath("//*[@id='main-content']//div[@class ='product']")
+        for href in product_links:
+            plink = href.xpath(".//div[@class='entry-media']/a/@href").extract()
+            full_url = response.urljoin(plink[0])
+            #availability of item
+            availabity = self.get_availability(href)
+            yield scrapy.Request(full_url, callback=self.parse_product, meta={'availabilty': availabity})
```
Comment (ali-ashraf-karimi) [inline gulahmed/gulahmed/spiders/gul_products.py]: availability instead of  availabity

File: gulahmed/gulahmed/spiders/gul_products.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
+__author__ = 'mateenahmeed'
+import scrapy
+from scrapy.spiders import CrawlSpider, Rule
+from scrapy.linkextractors.sgml import SgmlLinkExtractor
+from gulahmed.items import gulItem
+
+
+class MySpider(CrawlSpider):
```
Comment (ali-ashraf-karimi) [inline gulahmed/gulahmed/spiders/gul_products.py]: Better class name. (More meaningful)

File: gulahmed/gulahmed/spiders/gul_products.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+        # demanded as int  '3'
+        return 3
+
+    # for item code
+    def get_item_code(self, sel):
+        return sel.xpath("//*[@id='main-content']//header//h3/text()").extract()[0].strip()
+
+    # for product description
+    def get_description(self, sel):
+        des = sel.xpath("//*[@id='product-description']/p[2]/text()").extract()
```
Comment (ali-ashraf-karimi) [inline gulahmed/gulahmed/spiders/gul_products.py]: Avoid indexing in xpaths where possible
Try this: //*[@id='product-description']//p//text()

File: gulahmed/gulahmed/spiders/gul_products.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+        return sel.xpath("//*[@id='main-content']//header//h2/text()").extract()[0].strip()
+
+    # image url
+    def get_image_url(self, response):
+        src = response.xpath("//img[@id='zoom1']/@src")
+        return response.urljoin(src[0].extract())
+
+    # price for product
+    def get_price(self, sel):
+        price = sel.xpath("//article [@class='entry-content']//li[3]/span[@class= 'value']/text()").extract()[0].strip()
```
Comment (ali-ashraf-karimi) [inline gulahmed/gulahmed/spiders/gul_products.py]: Again indexing should be avoided. If they add or delete an li your xpath will break
Try: //article[@class='entry-content']//span[contains(., "Price")]/following::span[1]//text()

File: gulahmed/gulahmed/spiders/gul_products.py [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+        return price.split(" ")[-1]
+
+    def get_availability(self, sel):
+        availblilty = sel.xpath(".//div[@class='entry-main']/div/a/span/text()").extract()[0]
```
Comment (ali-ashraf-karimi) [inline gulahmed/gulahmed/spiders/gul_products.py]: Could be shortened to: //div[@class='entry-main']//span/text()

File: gulahmed/gulahmed/spiders/gul_products.py [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+    # for item code
+    def get_item_code(self, sel):
+        return sel.xpath("//*[@id='main-content']//header//h3/text()").extract()[0].strip()
+
+    # for product description
+    def get_description(self, sel):
+        des = sel.xpath("//*[@id='product-description']/p[2]/text()").extract()
+        if des:
+            return des[0]
+        return ""
```
Comment (ali-ashraf-karimi) [inline gulahmed/gulahmed/spiders/gul_products.py]: return des[0] if des else ""

File: gulahmed/gulahmed/spiders/gul_products.py [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+            plink = href.xpath(".//div[@class='entry-media']/a/@href").extract()
+            full_url = response.urljoin(plink[0])
+            #availability of item
+            availabity = self.get_availability(href)
+            yield scrapy.Request(full_url, callback=self.parse_product, meta={'availabilty': availabity})
+
+     # for each product
+    def parse_product(self, response):
+        sel = response.xpath("/html")
+        # item = response.meta["item"]
```
Comment (ali-ashraf-karimi) [inline gulahmed/gulahmed/spiders/gul_products.py]: Remove useless comments

File: gulahmed/gulahmed/items.py [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+    item_code= scrapy.Field()
+    item_price= scrapy.Field()
+    item_url= scrapy.Field()
+    item_image_url= scrapy.Field()
+    item_brand_id = scrapy.Field()
+    item_category_name= scrapy.Field()
+    item_description= scrapy.Field()
+    item_is_avaiable= scrapy.Field()
+    item_size = scrapy.Field()
+    item_color = scrapy.Field()
```
Comment (ali-ashraf-karimi) [inline gulahmed/gulahmed/items.py]: Remove unused variables

**Comment by ali-ashraf-karimi** [comment]
LGTM

File: gulahmed/gulahmed/items.py [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+class gulItem(scrapy.Item):
+    item_code= scrapy.Field()
+    item_price= scrapy.Field()
+    item_url= scrapy.Field()
+    item_image_url= scrapy.Field()
+    item_brand_id = scrapy.Field()
+    item_category_name= scrapy.Field()
+    item_description= scrapy.Field()
+    item_is_available= scrapy.Field()
+
```
Comment (alisufian) [inline gulahmed/gulahmed/items.py]: pep8 line spacing

File: gulahmed/gulahmed/items.py [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+# Define here the models for your scraped items
+#
+# See documentation in:
+# http://doc.scrapy.org/en/latest/topics/items.html
+
+import scrapy
+
+
+class gulItem(scrapy.Item):
```
Comment (alisufian) [inline gulahmed/gulahmed/items.py]: pep8 class name

Comment (alisufian) [inline gulahmed/gulahmed/items.py]: Not a good class name either. It should represent the item its going to hold.

File: gulahmed/gulahmed/spiders/gul_products.py [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+        Rule(SgmlLinkExtractor(restrict_xpaths=("//*[@id='site-menu']//ul[contains(@class,'mega-sub')]",))),
+        Rule(SgmlLinkExtractor(deny=("PriceCondition",), restrict_xpaths=("//*[@id='brands-list']",)),
+             callback="parse_items")
+    )
+
+    # this will parse all the products on current page
+    def parse_items(self, response):
+
+        sel = response.xpath("/html")
+        nextpage = sel.xpath("//a[@data-title='Next Page']/@href")
```
Comment (alisufian) [inline gulahmed/gulahmed/spiders/gul_products.py]: cant we add this to the rules ?

File: gulahmed/gulahmed/spiders/gul_products.py [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+        if nextpage:
+            next_page_url = response.urljoin(nextpage[0].extract())
+            yield scrapy.Request(next_page_url, callback=self.parse_items)
+
+        # getting all products on current page
+        product_links = response.xpath("//*[@id='main-content']//div[@class ='product']")
```
Comment (alisufian) [inline gulahmed/gulahmed/spiders/gul_products.py]: this should also go into the rules. No need to parse manually.

File: gulahmed/gulahmed/spiders/gul_products.py [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+        return response.urljoin(src[0].extract())
+
+    # price for product
+    def get_price(self, sel):
+        price = sel.xpath("//article[@class='entry-content']"
+                          "//span[contains(., 'Price')]/following::span[1]//text()").extract()[0].strip()
+        return price.split(" ")[-1]
+
+    def get_availability(self, sel):
+        availblilty = sel.xpath(".//div[@class='entry-main']//span/text()").extract()[0]
```
Comment (alisufian) [inline gulahmed/gulahmed/spiders/gul_products.py]: wrong spellings `availblilty`

File: gulahmed/gulahmed/spiders/gul_products.py [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+        if "Not" in availblilty:
+            return False
+        return True
```
Comment (alisufian) [inline gulahmed/gulahmed/spiders/gul_products.py]: ``return not "Not" in availblilty`

File: gulahmed/gulahmed/spiders/gul_products.py [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+        item['item_category_name'] = self.get_category(sel)
+        item['item_brand_id'] = self.get_brand_id(sel)
+        item['item_code'] = self.get_item_code(sel)
+        item['item_image_url'] = self.get_image_url(response)
+        item['item_description'] = self.get_description(sel)
+        item['item_price'] = self.get_price(sel)
+        yield item
+
+    # for brand id
+    def get_brand_id(self, sel):
```
Comment (alisufian) [inline gulahmed/gulahmed/spiders/gul_products.py]: brand_id should be a class level variable.
base class of the spider should set it.

File: kaprescrapers/kaprescrapers/items.py [hunk 18]
```diff
@@ -0,0 +1,20 @@
...
+
+# Define here the models for your scraped items
+#
+# See documentation in:
+# http://doc.scrapy.org/en/latest/topics/items.html
+
+import scrapy
+
+
+class KaprePkItem(scrapy.Item):
```
Comment (alisufian) [inline kaprescrapers/kaprescrapers/items.py]: Lets call this Garment since thats what it is.

**Comment by alisufian** [comment]
Its always a good idea to create your own base class instead of using the 3rd party's directly.
```python
class KapreBaseSpider(CrawlSpider):
   ........

class GulAhmedSpider(KapreBaseSpider):
   .......
```

File: kaprescrapers/kaprescrapers/spiders/alkaram_spider.py [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+
+    # for item code
+    def get_item_code(self, sel):
+        return sel.xpath(".//*[@class='product-name']/h1/text()").extract()[0].strip()
+
+    # for product description
+    def get_description(self, sel):
+        des = sel.xpath(".//tbody/tr[contains(.,'Description')]//td/p/text()").extract()
+        if not des:
+            des = sel.xpath(".//tbody/tr[contains(.,'Description')]//td/text()").extract()
```
Comment (alisufian) [inline kaprescrapers/kaprescrapers/spiders/alkaram_spider.py]: Why cant we do `sel.xpath(".//tbody/tr[contains(.,'Description')]//td//text()")`

File: kaprescrapers/kaprescrapers/spiders/alkaram_spider.py [hunk 20]
```diff
@@ -0,0 +1,70 @@
...
+    # image url
+    def get_image_url(self, response):
+        src = response.xpath(".//*[@id='image-zoom']/@href")
+        return response.urljoin(src[0].extract())
+
+    # price for product
+    def get_price(self, sel):
+        price = sel.xpath(".//*[@class='regular-price']//*[@class='price']/text()").extract()
+        if not price:
+            price = sel.xpath(".//*[@class='special-price']//*[@class='price']/text()").extract()
```
Comment (alisufian) [inline kaprescrapers/kaprescrapers/spiders/alkaram_spider.py]: `sel.xpath(".//*[@class='regular-price' or @class='special-price']//*[@class='price']/text()").extract()`

File: kaprescrapers/kaprescrapers/spiders/alkaram_spider.py [hunk 21]
```diff
@@ -0,0 +1,70 @@
...
+        return (price[0].strip()).split(" ")[-1]
```
Comment (alisufian) [inline kaprescrapers/kaprescrapers/spiders/alkaram_spider.py]: Any better way to do this?

File: kaprescrapers/kaprescrapers/spiders/bareeze_spider.py [hunk 22]
```diff
@@ -0,0 +1,80 @@
...
+
+    # image url
+    def get_image_url(self, response):
+        src = response.xpath("//*[@id='cloudZoom']/@href")
+        return response.urljoin(src[0].extract())
+
+    # price for product
+    def get_price(self, sel):
+        price = sel.xpath(".//*[@class= 'price']/text()").extract()[-1].strip()
+        return price.split(" ")[-1]
```
Comment (alisufian) [inline kaprescrapers/kaprescrapers/spiders/bareeze_spider.py]: You dont need to pass parameters to split on whitespace
`price.split()[-1]`

File: kaprescrapers/kaprescrapers/spiders/gulahmed_spider.py [hunk 23]
```diff
@@ -0,0 +1,75 @@
...
+        Rule(SgmlLinkExtractor(deny=("PriceCondition",), restrict_xpaths=("//*[@id='brands-list']",
+                                                                          "//a[@data-title='Next Page']")),
+             follow=True,
+             callback="parse_items")
+    )
+
+    # this will parse all the products on current page
+    def parse_items(self, response):
+        # getting all products on current page
+        product_links = response.xpath("//*[@id='main-content']//div[@class ='product']")
```
Comment (alisufian) [inline kaprescrapers/kaprescrapers/spiders/gulahmed_spider.py]: extra spacing inside xpath

File: kaprescrapers/kaprescrapers/settings.py [hunk 24]
```diff
@@ -0,0 +1,85 @@
...
+#     http://scrapy.readthedocs.org/en/latest/topics/spider-middleware.html
+
+BOT_NAME = 'kaprescrapers'
+
+SPIDER_MODULES = ['kaprescrapers.spiders']
+NEWSPIDER_MODULE = 'kaprescrapers.spiders'
+
+
+# Crawl responsibly by identifying yourself (and your website) on the user-agent
+#USER_AGENT = 'kaprescrapers (+http://www.yourdomain.com)'
```
Comment (alisufian) [inline kaprescrapers/kaprescrapers/settings.py]: Use a browser user agent

File: kaprescrapers/kaprescrapers/settings.py [hunk 25]
```diff
@@ -0,0 +1,85 @@
...
+# Configure item pipelines
+# See http://scrapy.readthedocs.org/en/latest/topics/item-pipeline.html
+#ITEM_PIPELINES = {
+#    'kaprescrapers.pipelines.SomePipeline': 300,
+#}
+
+# Enable and configure the AutoThrottle extension (disabled by default)
+# See http://doc.scrapy.org/en/latest/topics/autothrottle.html
+# NOTE: AutoThrottle will honour the standard settings for concurrency and delay
+#AUTOTHROTTLE_ENABLED=True
```
Comment (alisufian) [inline kaprescrapers/kaprescrapers/settings.py]: enable this

**Pull request merged**